tests/test_admin.py

```diff
@@ -0,0 +1,167 @@
+        self.assertEqual(set(site1_pagecontents), set(response1.context["cl"].queryset))
+        self.assertEqual(set(site2_pagecontents), set(response2.context["cl"].queryset))
+
+    def test_language_filter(self):
+        expected_en = PageContentWithVersionFactory.create_batch(3, language="en")
+        expected_de = PageContentWithVersionFactory.create_batch(3, language="de")
+        model = PageContent
+        base_url = self.get_admin_url(model, "changelist")
+        with self.login_user_context(self.get_superuser()):
+            # en is the default language configured for the site
+            response_default = self.client.get(base_url)
+            qs_default = response_default.context["cl"].queryset
+            response_en = self.client.get(base_url + "?language=en")
+            qs_en = response_en.context["cl"].queryset
+            response_de = self.client.get(base_url + "?language=de")
+            qs_de = response_de.context["cl"].queryset
+
+        self.assertEqual(set(qs_default), set(expected_en))
+        self.assertEqual(set(qs_en), set(expected_en))
+        self.assertEqual(set(qs_de), set(expected_de))
+
+    def test_unpublished_filter(self):
+        expected = PageContentWithVersionFactory.create_batch(3, language="en")
+        expected_unpublished = PageContentWithVersionFactory.create_batch(
+            2, language="en", version__state=UNPUBLISHED
+        )
+        model = PageContent
+        base_url = self.get_admin_url(model, "changelist")
+        with self.login_user_context(self.get_superuser()):
+            # en is the default language configured for the site
+            response_default = self.client.get(base_url)
+            qs_default = response_default.context["cl"].queryset
+            response_unpublished = self.client.get(base_url + "?unpublished=1")
+            qs_unpublished = response_unpublished.context["cl"].queryset
+
+        self.assertEqual(set(qs_default), set(expected))
+        self.assertEqual(set(qs_unpublished), set(expected_unpublished))
+
+
+class DuplicateViewTestCase(CMSTestCase):
+    def test_obj_does_not_exist(self):
+        with self.login_user_context(self.get_superuser()), patch(
+            "djangocms_pageadmin.admin.PageContentAdmin._get_obj_does_not_exist_redirect",
+            return_value=HttpResponse(),
+        ) as mock:
+            response = self.client.get(
+                self.get_admin_url(PageContent, "duplicate_content", "foo")
+            )
+        mock.assert_called_once()
+        self.assertEqual(response, mock.return_value)
+
+    def test_get(self):
+        pagecontent = PageContentWithVersionFactory()
+        with self.login_user_context(self.get_superuser()):
+            response = self.client.get(
+                self.get_admin_url(PageContent, "duplicate_content", pagecontent.pk)
+            )
+        self.assertEqual(response.status_code, 200)
+
+    def test_post_empty_slug(self):
```

Comment:
In this case the validation is simple enough that it's probably fine to do as an integration test. But if the validation ever gets more complicated it would probably be better to do this as a set of unit tests (for one thing they tend to be a tiny bit faster).